src/training/graph_group.h

```diff
@@ -112,12 +112,12 @@ class AsyncGraphGroup : public GraphGroup {
           graph->forward();
         }
 
-        if(shardOpt_.size() == 0){
-          for (auto device : devices_) {
+        if(shardOpt_.size() == 0)
+          for (auto device : devices_) 
             shardOpt_.push_back(Optimizer(options_));
 
-          }
-        }
+        shardSync_.reset( new std::mutex[devices_.size()] );
+         
```

Comment:
Does unique_ptr know how to handle arrays? std::mutex is probably not copyable, but you should be able to create a vector of mutex with emplace and other move contructor functionality.